packages/components/src/number-control/index.js

```diff
@@ -180,10 +179,7 @@ export function NumberControl(
 			step={ step }
 			type={ typeProp }
 			value={ valueProp }
-			__unstableStateReducer={ composeStateReducers(
-				numberControlStateReducer,
-				stateReducer
-			) }
+			__unstableStateReducer={ numberControlStateReducer }
```

Comment:
I do generally prefer the prop `stateReducer` to be composed at this point in the code, rather than be directly inserted into the `numberControlStateReducer` function. I think it's easier for a casual code reader to understand the intent at first glance. Might just be me though.

```suggestion
			__unstableStateReducer={ compose( stateReducer, numberControlStateReducer ) }
```

(Using `wp.data.compose`)

Reply:
This is actually a good point

Reply:
The same thought had occurred to me about visibility and, reading your suggestion, I had the gut reaction:
"That looks great! lets do it!"
Then I tried it and realized it's not that simple. It was a interesting pursuit 😄. If I've missed something let me know.
>(Using `wp.data.compose`)

<details><summary>I figured you meant `compose` from the package of the same name. Using that for composing reducers requires intermediary/wrapper functions to port the two arguments into any secondary reducer.</summary>

```javascript
			__unstableStateReducer={ 
				compose( [
					( args ) => stateReducer( ...args ),
					( state, action ) => [
						numberControlStateReducer( state, action ),
						action,
					]
				] )
			}
```
</details>

<details><summary>So I dug around and realized maybe you meant `combineReducers` from the data package. That looks nice but creates nested state objects so needs a final pass to merge them.
</summary>

```javascript
	let usedStateReducer = numberControlStateReducer;
	if ( stateReducer ) {
		const combinedReducer = combineReducers( {
			outer: stateReducer,
			inner: numberControlStateReducer,
		} );
		usedStateReducer = ( state, action ) => {
			const result = combinedStateReducer( state, action );
			return { ...result.outer, ...result.inner };
		};
	}
```
</details>

<details><summary>Employing either `compose` or `combineReducers` here gets a bit verbose and begs to be generalized into a replacement for`composeStateReducers`. But a dynamic compose-like utility is still overkill for the established pattern in which no more than two are ever combined at once. If we must have a helper my vote would be for something like `pairStateReducers( b, a )`</summary>

```javascript
const pairStateReducers = ( outer, inner ) => ( state, action ) => {
	const baseState = inner( state, action );
	return outer?.( baseState, action ) ?? baseState;
};
```
</details>

Directly coupling each pair may be a bit verbose but it's plain to see which one runs before the other. For now, I've stuck with that approach but moved the coupling to outside the reducers themselves.

Reply:
This exploration makes sense, even more so when thinking that [`compose` is simply a re-export of `flowRight`](https://github.com/WordPress/gutenberg/blob/a3fa82c3ec0a57f2f09d438a36d2ae86ff805b03/packages/compose/src/higher-order/compose.ts) and that the original `composeStateReducers` function that is being deleted, already worked by using `reduceRight`.

We can wait for @mirka 's comment, but it seems to me that we can keep the current approach and merge this PR

Reply:
Not as simple as it seemed, huh! Thanks for the great summary of your investigation, @stokesman.

Agreed, I don't think it's a blocker either, no strong need for an abstraction at this stage. Let's merge 🚢